tower-grpc-interop/src/client.rs

```diff
@@ -278,33 +318,202 @@ impl Testcase {
+                                test_assert!(
+                                    "call must be successful",
+                                    result.is_ok(),
+                                    format!("result={:?}", result)
+                                )
+                        ];
+                        if let Ok(response) = result.map(|r| r.into_inner()) {
+                            assertions.push(test_assert!(
+                            "aggregated payload size must be 74922 bytes",
+                            response.aggregated_payload_size == 74922,
+                            format!("aggregated_payload_size={:?}",
+                                response.aggregated_payload_size
+                            )));
+                        }
+                        future::ok::<Vec<TestAssertion>, Box<Error>>(assertions)
+                    }))
+            },
+            Testcase::server_streaming => {
+                use pb::ResponseParameters;
+                let req = pb::StreamingOutputCallRequest {
+                    response_parameters: RESPONSE_LENGTHS.iter().map(|len| {
+                        ResponseParameters::with_size(*len)
+                    }).collect(),
+                    ..Default::default()
+                };
+                let req = Request::new(req);
+                core.run(client
+                    .streaming_output_call(req)
+                    .map_err(|tower_err: tower_grpc::Error<tower_h2::client::Error>| -> Box<Error> {
+                        Box::new(tower_err)
+                    })
+                    .and_then(|response_stream| {
+                        // Convert the stream into a plain Vec
+                        response_stream.into_inner()
+                            .collect()
+                            .map_err(|tower_err: tower_grpc::Error<()>| -> Box<Error> {
+                                Box::new(tower_err)
+                            })
+                    })
+                    .map(|responses: Vec<pb::StreamingOutputCallResponse>| -> Vec<TestAssertion> {
+                        vec![
+                            test_assert!(
+                                "there should be four responses",
+                                responses.len() == 4,
+                                format!("responses.len()={:?}", responses.len())
+                            ),
+                            test_assert!(
+                                "the response payload sizes should match input",
+                                RESPONSE_LENGTHS == response_lengths(&responses).as_slice(),
+                                format!("{:?}={:?}", RESPONSE_LENGTHS, response_lengths(&responses))
+                            ),
+                        ]
+                    })
+                    .then(&assert_success))
+            },
+            Testcase::ping_pong => {
+                fn make_req(idx: usize) -> pb::StreamingOutputCallRequest {
```

Comment:
Personally, I think it would be easier to read this code if the helper functions/types for doing the ping-pong weren't local to this function --- we're already indented several levels here. Take it or leave it.